crates/ibc/src/core/ics24_host/identifier.rs

```diff
@@ -131,19 +117,38 @@ impl ChainId {
     /// assert_eq!(ChainId::new("chainA", 1).with_version(2), ChainId::new("chainA", 2));
     /// assert_eq!("chain1".parse::<ChainId>().unwrap().with_version(2), "chain1".parse::<ChainId>().unwrap());
     /// ```
-    pub fn with_version(mut self, version: u64) -> Self {
-        if Self::is_epoch_format(&self.id) {
-            self.id = {
-                let mut split: Vec<&str> = self.id.split('-').collect();
-                let version = version.to_string();
-                if let Some(last_elem) = split.last_mut() {
-                    *last_elem = &version;
-                }
-                split.join("-")
-            };
-            self.version = version;
+    pub fn with_version(self, version: u64) -> Self {
+        if self.version == 0 {
+            self
+        } else {
+            let (name, _) = Self::split_chain_id(&self.id).unwrap();
+            // TODO: We're wasting allocation here.  We own self.id, so we could
+            // `self.id.truncate(name.len() + 1); write!(&mut self.id, "{}",
+            // version)`. Unfortunately, we’re no_std so `write!` is out and
+            // `self.id.push_str(&version.to_string());` defeats the purpose of
+            // the optimisation.  `itoa` crate is possible option.
+            Self::new(name, version)
+        }
+    }
+
+    /// Splits chain_id into name and version if the id includes epoch number.
+    ///
+    /// Chain id with epoch number has format `{name}-{version}` where name is
+    /// non-empty and version is a positive integer starting with a non-zero
+    /// digit.  If `chain_id` is in that format, returns `(name, version)`;
+    /// otherwise returns an error.
+    fn split_chain_id(chain_id: &str) -> Result<(&str, &str), ()> {
+        let (name, version) = chain_id.rsplit_once('-').ok_or(())?;
+        // `-123` and `foo--123` are not in epoch format.
+        if name.is_empty() || name.as_bytes().last() == Some(&b'-') {
+            return Err(());
+        }
+        let (car, cdr) = version.as_bytes().split_first().ok_or(())?;
+        if matches!(*car, b'1'..=b'9') && cdr.iter().all(u8::is_ascii_digit) {
+            Ok((name, version))
+        } else {
+            Err(())
         }
-        self
     }
```

Comment:
This function works independent of any instance of `ChainId`. Good to promote its usability and having it as a free function, imo.